src/client/datascience/kernel-launcher/kernelEnvVarsService.ts

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import { inject, injectable } from 'inversify';
+import { traceError } from '../../common/logger';
+import { IPlatformService } from '../../common/platform/types';
+import { Resource } from '../../common/types';
+import { IEnvironmentVariablesService } from '../../common/variables/types';
+import { IEnvironmentActivationService } from '../../interpreter/activation/types';
+import { IInterpreterService } from '../../interpreter/contracts';
+import { EnvironmentType } from '../../pythonEnvironments/info';
+import { IJupyterKernelSpec } from '../types';
+
+@injectable()
+export class KernelEnvironmentVariablesService {
+    constructor(
+        @inject(IInterpreterService) private readonly interpreterService: IInterpreterService,
+        @inject(IEnvironmentActivationService) private readonly envActivation: IEnvironmentActivationService,
+        @inject(IEnvironmentVariablesService) private readonly envVarsService: IEnvironmentVariablesService,
+        @inject(IPlatformService) private readonly platformService: IPlatformService
+    ) {}
+    /**
+     * If the kernel belongs to a conda environment, then use the env variables of the conda environment and merge that with the env variables of the kernel spec.
+     * In the case of some kernels such as java, the kernel spec contains the cli as such `argv = ['java', 'xyz']`.
+     * The first argument is an executable, and it is not in the current path.
+     * However, when activating the conda env, the path variables are updated to set path to the location where the java executable is located.
+     */
+    public async getEnvironmentVariables(resource: Resource, kernelSpec: IJupyterKernelSpec) {
+        let kernelEnv = kernelSpec.env && Object.keys(kernelSpec.env).length > 0 ? kernelSpec.env : undefined;
+        if (!kernelSpec.interpreterPath) {
+            return kernelEnv;
+        }
+        const interpreter = await this.interpreterService
+            .getInterpreterDetails(kernelSpec.interpreterPath)
+            .catch((ex) => {
+                traceError('Failed to fetch interpreter information for interpreter that owns a kernel', ex);
+                return undefined;
+            });
+
+        if (interpreter?.envType !== EnvironmentType.Conda) {
+            return kernelEnv;
+        }
+        const interpreterEnv = await this.envActivation.getActivatedEnvironmentVariables(resource, interpreter, true);
+        if (!interpreterEnv) {
+            return kernelEnv;
+        }
+
+        // Merge the env variables with that of the kernel env.
+        const mergedVars = { ...process.env };
+        kernelEnv = kernelEnv || {};
+        this.envVarsService.mergeVariables(interpreterEnv, mergedVars);
```

Comment:
Looking at the merge function. Is this correct if interpreter or kernel specifies something already in the process.env? Trying to look it up now, but I would think that it would be appropriate for the kernelEnv to win out over the process env. But here it looks like the process env value would win.

Reply:
Oops, you are absolutely right.
I remember keeping that in mind when i did this...lol

Reply:
Yes, this is correct

Reply:
if you look, like 52, tha'ts where we apply the variables from `kernelenv` last, hence that wins.
I.e. variables from `kernelEnv` is applied over the rest.
**Happy to improve the API (signature or order of parameters) to make it more obvious.**

Reply:
Should i change the method from `mergeVariabels` into `overrideEnvVariables`, still not sure that helps.

Reply:
Hmm, give me a second. Maybe I'm reading this wrong. 

What is see is that in mergeVariables it's (source, target) and in the function target[KEY] is only replaced by source[KEY] if target[KEY] is undefined. 

mergedVars starts with process.env so if a key is specified there it won't get replaced by interpreter or kernel as merged is always the target, right?

Reply:
Yikes, yes. Looks like a bug. will need to review that code & its usage.
Thanks.

Reply:
Thanks fixed to remove the check for `undefined` & flipped the arguments in the other one place where this was used.

Reply:
@DonJayamanne Thanks, that seems like a good fix. Merge functions can be funky in the "who wins a tie" scenario. If you are still making the fix could stick a comment on the mergeVariables function just to make it clear how the merge is happening. Moving my review to approve since this was the only comment I was still worried about.

Reply:
Agreed, done.
Please re-review if possible. Added comments